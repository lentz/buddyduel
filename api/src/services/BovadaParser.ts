import * as jp from 'jsonpath';
import { inspect } from 'util';
import IGame from '../models/IGame';

const teamRegex = /^[\w\s\d()#&.]+$/;

function parseGame(game: IGame): IGame | null {
  let pointSpread;
  try {
    [pointSpread] = jp.query(
      game,
      '$..markets[?(@.description=="Point Spread" && @.status=="O")]',
    );
    if (!pointSpread) {
      return null;
    }
    const home = pointSpread.outcomes.find(
      (team: { type: string }) => team.type === 'H',
    );
    const away = pointSpread.outcomes.find(
      (team: { type: string }) => team.type === 'A',
    );
    if (
      !home?.description?.match(teamRegex) ||
      !away?.description?.match(teamRegex)
    ) {
      return null;
    }
    return {
      id: game.id,
      homeTeam: home.description.trim(),
      homeSpread: Number(home.price.handicap),
      awayTeam: away.description.trim(),
      awaySpread: Number(away.price.handicap),
      startTime: game.startTime,
    };
  } catch (err) {
    console.warn(
      `Error parsing Bovada game: ${err}\nPointSpread JSON: ${inspect(
        pointSpread,
      )}`,
    ); // eslint-disable-line no-console
    return null;
  }
}

export function parseGames(json: any) {
  return jp
    .query(json, '$..events[*]')
    .map(parseGame)
    .filter((game) => game) as IGame[];
}
