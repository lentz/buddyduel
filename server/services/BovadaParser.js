const jp = require('jsonpath');
const util = require('util');

const teamRegex = /^[\w\s\d()#&.]+$/;

function parseGame(game) {
  let pointSpread;
  try {
    [pointSpread] = jp.query(
      game,
      '$..markets[?(@.description=="Point Spread" && @.status=="O")]',
    );
    if (!pointSpread) { return null; }
    const home = pointSpread.outcomes.find(team => team.type === 'H');
    const away = pointSpread.outcomes.find(team => team.type === 'A');
    if (!home.description.match(teamRegex) || !away.description.match(teamRegex)) {
      return null;
    }
    return {
      id: game.id,
      homeTeam: home.description.trim(),
      homeSpread: Number(home.price.handicap),
      awayTeam: away.description.trim(),
      awaySpread: Number(away.price.handicap),
      startTime: game.startTime,
    };
  } catch (err) {
    console.warn(`Error parsing Bovada game: ${err}\nPointSpread JSON: ${util.inspect(pointSpread)}`); // eslint-disable-line no-console
    return null;
  }
}

module.exports.parseGames = function parseGames(json) {
  return jp
    .query(json, '$..events[*]')
    .map(parseGame)
    .filter(game => game);
};
