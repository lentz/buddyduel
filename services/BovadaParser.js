const crypto = require('crypto');
const jp = require('jsonpath');
const NFLWeek = require('./NFLWeek');
const util = require('util');

const teamRegex = /^[\w\s\d]+$/;

function parseGame(game) {
  let pointSpread;
  try {
    pointSpread = jp.query(game, '$..itemList[?(@.description=="Point Spread")]')[0];
    if (!pointSpread) { return null; }
    const home = pointSpread.outcomes.find(team => team.type === 'H');
    const away = pointSpread.outcomes.find(team => team.type === 'A');
    if (!home.description.match(teamRegex) || !away.description.match(teamRegex)) {
      return null;
    }
    return {
      id: crypto.createHash('md5')
        .update(`${home.description}|${away.description}|${NFLWeek.seasonYear}|${NFLWeek.forGame(game)}`)
        .digest('hex'),
      homeTeam: home.description,
      homeSpread: Number(home.price.handicap),
      awayTeam: away.description,
      awaySpread: Number(away.price.handicap),
      startTime: game.startTime,
    };
  } catch (err) {
    console.warn(`Error parsing Bovada game: ${err}\nPointSpread JSON: ${util.inspect(pointSpread)}`); // eslint-disable-line no-console
    return null;
  }
}

module.exports.call = json => jp
  .query(json, '$..items[?(@.type=="NFL" || @.type=="NFL, Playoffs")]')
  .map(parseGame)
  .filter(game => game);
