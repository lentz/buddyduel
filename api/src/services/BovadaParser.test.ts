import { readFileSync } from 'fs';
import { join } from 'path';
import * as BovadaParser from './BovadaParser';

describe('BovadaParser', () => {
  describe('#parseGames()', () => {
    const bovadaJSON = JSON.parse(readFileSync(join(
      __dirname, '..', '..', '__tests__', 'sample-data', 'bovada-nfl.json',
    )).toString());

    test('parses the JSON and returns game objects', () => {
      expect(BovadaParser.parseGames(bovadaJSON)).toMatchSnapshot();
    });
  });
});
