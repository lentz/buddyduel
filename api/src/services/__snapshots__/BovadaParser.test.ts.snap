// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BovadaParser #parseGames() parses the JSON and returns game objects 1`] = `
Array [
  Object {
    "awaySpread": 4,
    "awayTeam": "Atlanta Falcons",
    "homeSpread": -4,
    "homeTeam": "Philadelphia Eagles",
    "id": "3172275",
    "startTime": 1536279600000,
  },
  Object {
    "awaySpread": 7,
    "awayTeam": "Houston Texans",
    "homeSpread": -7,
    "homeTeam": "New England Patriots",
    "id": "3172285",
    "startTime": 1536512700000,
  },
  Object {
    "awaySpread": 5.5,
    "awayTeam": "San Francisco 49ers",
    "homeSpread": -5.5,
    "homeTeam": "Minnesota Vikings",
    "id": "3172278",
    "startTime": 1536512700000,
  },
  Object {
    "awaySpread": -3.5,
    "awayTeam": "Jacksonville Jaguars",
    "homeSpread": 3.5,
    "homeTeam": "New York Giants",
    "id": "3172283",
    "startTime": 1536512700000,
  },
  Object {
    "awaySpread": 2.5,
    "awayTeam": "Cincinnati Bengals",
    "homeSpread": -2.5,
    "homeTeam": "Indianapolis Colts",
    "id": "3172280",
    "startTime": 1536512700000,
  },
  Object {
    "awaySpread": -6,
    "awayTeam": "Pittsburgh Steelers",
    "homeSpread": 6,
    "homeTeam": "Cleveland Browns",
    "id": "3172276",
    "startTime": 1536512700000,
  },
  Object {
    "awaySpread": -2,
    "awayTeam": "Tennessee Titans",
    "homeSpread": 2,
    "homeTeam": "Miami Dolphins",
    "id": "3172286",
    "startTime": 1536512700000,
  },
  Object {
    "awaySpread": 3,
    "awayTeam": "Kansas City Chiefs",
    "homeSpread": -3,
    "homeTeam": "Los Angeles Chargers",
    "id": "3172289",
    "startTime": 1536523500000,
  },
  Object {
    "awaySpread": 2,
    "awayTeam": "Seattle Seahawks",
    "homeSpread": -2,
    "homeTeam": "Denver Broncos",
    "id": "3172290",
    "startTime": 1536524700000,
  },
  Object {
    "awaySpread": 2.5,
    "awayTeam": "Dallas Cowboys",
    "homeSpread": -2.5,
    "homeTeam": "Carolina Panthers",
    "id": "3172291",
    "startTime": 1536524700000,
  },
  Object {
    "awaySpread": 0,
    "awayTeam": "Washington Redskins",
    "homeSpread": 0,
    "homeTeam": "Arizona Cardinals",
    "id": "3172292",
    "startTime": 1536524700000,
  },
  Object {
    "awaySpread": 8,
    "awayTeam": "Chicago Bears",
    "homeSpread": -8,
    "homeTeam": "Green Bay Packers",
    "id": "3172294",
    "startTime": 1536538800000,
  },
  Object {
    "awaySpread": 7,
    "awayTeam": "New York Jets",
    "homeSpread": -7,
    "homeTeam": "Detroit Lions",
    "id": "3172296",
    "startTime": 1536621000000,
  },
  Object {
    "awaySpread": -3,
    "awayTeam": "Los Angeles Rams",
    "homeSpread": 3,
    "homeTeam": "Oakland Raiders",
    "id": "3172305",
    "startTime": 1536632400000,
  },
]
`;
